src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
 pub mod trust_store;
 pub mod types;
 
+use std::collections::HashSet;
+use std::vec;
+
+use crate::certificate::cert_is_self_issued;
+use crate::types::{DNSConstraint, IPAddress, IPConstraint};
+use crate::ApplyNameConstraintStatus::{Applied, Skipped};
+use cryptography_x509::extensions::{DuplicateExtensionsError, Extensions};
+use cryptography_x509::{
+    certificate::Certificate,
+    extensions::{NameConstraints, SubjectAlternativeName},
+    name::GeneralName,
+    oid::{NAME_CONSTRAINTS_OID, SUBJECT_ALTERNATIVE_NAME_OID},
+};
+use ops::CryptoOps;
+use policy::Policy;
+use trust_store::Store;
+use types::DNSName;
+
+#[derive(Debug, PartialEq, Eq)]
 pub enum ValidationError {
-    Other(&'static str),
+    CandidatesExhausted(Box<ValidationError>),
+    Malformed(asn1::ParseError),
+    DuplicateExtension(DuplicateExtensionsError),
+    Other(String),
+}
+
+impl From<asn1::ParseError> for ValidationError {
+    fn from(value: asn1::ParseError) -> Self {
+        Self::Malformed(value)
+    }
+}
+
+impl From<DuplicateExtensionsError> for ValidationError {
+    fn from(value: DuplicateExtensionsError) -> Self {
+        Self::DuplicateExtension(value)
+    }
+}
+
+#[derive(Default)]
+struct AccumulatedNameConstraints<'a> {
+    sans: Vec<GeneralName<'a>>,
+    name_constraints: Vec<NameConstraints<'a>>,
+}
+
+impl<'a> AccumulatedNameConstraints<'a> {
+    fn accumulate_san(&mut self, extensions: &Extensions<'a>) -> Result<(), ValidationError> {
+        if let Some(sans) = extensions.get_extension(&SUBJECT_ALTERNATIVE_NAME_OID) {
+            let sans: SubjectAlternativeName<'_> = sans.value()?;
+            self.sans.extend(sans);
+        }
+
+        Ok(())
+    }
+
+    fn apply_inner(
+        &self,
+        constraint: &GeneralName<'a>,
+        san: &GeneralName<'_>,
+    ) -> Result<ApplyNameConstraintStatus, ValidationError> {
+        match (constraint, san) {
+            (GeneralName::DNSName(pattern), GeneralName::DNSName(name)) => {
+                match (DNSConstraint::new(pattern.0), DNSName::new(name.0)) {
+                    (Some(pattern), Some(name)) => Ok(Applied(pattern.matches(&name))),
+                    (Some(_), None) => Err(ValidationError::Other(format!(
+                        "unsatisfiable DNS name constraint: NC {} cannot match SAN {}",
+                        pattern.0, name.0
+                    ))),
```

Comment:
This isn't about the NC can't match the SAN, it's about the SAN not containing a valid DNS Name, isn't it?

Reply:
Yeah, this is a funny phrasing: it's meant to communicate that the NC cannot possibly match the SAN because it's malformed. I'll clarify.